cadc-storage-adapter-ad/src/main/java/org/opencadc/inventory/storage/ad/AdStorageAdapter.java

```diff
@@ -0,0 +1,332 @@
+import ca.nrc.cadc.io.WriteException;
+
+import ca.nrc.cadc.net.HttpDownload;
+import ca.nrc.cadc.net.IncorrectContentChecksumException;
+import ca.nrc.cadc.net.IncorrectContentLengthException;
+import ca.nrc.cadc.net.ResourceNotFoundException;
+import ca.nrc.cadc.net.TransientException;
+import ca.nrc.cadc.reg.Capabilities;
+import ca.nrc.cadc.reg.Capability;
+import ca.nrc.cadc.reg.Interface;
+import ca.nrc.cadc.reg.Standards;
+import ca.nrc.cadc.reg.client.RegistryClient;
+import ca.nrc.cadc.util.StringUtil;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URL;
+import java.security.AccessControlException;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.SortedSet;
+
+import javax.security.auth.Subject;
+
+import org.apache.log4j.Logger;
+import org.opencadc.inventory.InventoryUtil;
+import org.opencadc.inventory.StorageLocation;
+import org.opencadc.inventory.storage.NewArtifact;
+import org.opencadc.inventory.storage.StorageAdapter;
+import org.opencadc.inventory.storage.StorageEngageException;
+import org.opencadc.inventory.storage.StorageMetadata;
+import org.opencadc.tap.TapClient;
+
+/**
+ * The interface to storage implementations.
+ * 
+ * @author jeevesh
+ *
+ */
+public class AdStorageAdapter implements StorageAdapter {
+
+    private static final Logger log = Logger.getLogger(AdStorageAdapter.class);
+    private static final URI DATA_RESOURCE_ID = URI.create("ivo://cadc.nrc.ca/data");
+    private static final String UNABLE_TO_GET = "Unable to get from source: ";
+    private static final String TAP_SERVICE_URI = "ivo://cadc.nrc.ca/ad";
+
+    /**
+     * Construct an AdStorageAdapter with the config stored in the
+     * well-known properties file with well-known properties.
+     */
+    public AdStorageAdapter(){}
+
+    /**
+     * Get from storage the artifact identified by storageLocation.
+     * 
+     * @param storageLocation The storage location containing storageID and storageBucket.
+     * @param dest The destination stream.
+     * 
+     * @throws ResourceNotFoundException If the artifact could not be found.
+     * @throws ReadException If the storage system failed to stream.
+     * @throws WriteException If the client failed to stream.
+     * @throws StorageEngageException If the adapter failed to interact with storage.
+     * @throws TransientException If an unexpected, temporary exception occurred. 
+     */
+    public void get(StorageLocation storageLocation, OutputStream dest)
+        throws ResourceNotFoundException, ReadException, WriteException, StorageEngageException, TransientException {
+
+        URL sourceURL = this.toURL(storageLocation.getStorageID());
+        log.debug("sourceURL: " + sourceURL.toString());
+
+        HttpDownload get = new HttpDownload(sourceURL, dest);
+        get.run();
+
```

Comment:
HttpDownload already converts every 4xx and 5xx status code to the relevant exception and correctly gets the error message from the server. The hard part is converting the existing Throwable into an exception to throw directly. Unfortunately, I didn't want to fully adapt HttpDownload to support prepare() but in the long run that is the right way to do this... then this code can use try/catch instead of if (get.getThrowable() instanceof ...) or re-coding the response code to exception logic.

I will take a stab at making HttpDownload.prepare() work when I get home...

Reply:
cadc-util.1.3.1 includes a class HttpGet that supports prepare() and getInputStream(). The prepare() method throws all the exceptions and then the InputStream can be passed to an instanceof ThreadedIO to perform the IO loop and possibly throw some more exceptions.

With this, you'll only have to worry about exceptions that you can do something about. If you think that the get method in the StorageAdapter interface needs to declare more throws: that's possible/likely and we can discuss which are suitable.

Reply:
reworked